roles/tests-integration/tests/common/sniffer.rs

```diff
@@ -0,0 +1,403 @@
+            },
+            AnyMessage::TemplateDistribution(m) => match m {
+                CoinbaseOutputDataSize(m) => {
+                    AnyMessage::TemplateDistribution(CoinbaseOutputDataSize(m.into_static()))
+                }
+                TemplateDistribution::NewTemplate(m) => AnyMessage::TemplateDistribution(
+                    TemplateDistribution::NewTemplate(m.into_static()),
+                ),
+                TemplateDistribution::RequestTransactionData(m) => {
+                    AnyMessage::TemplateDistribution(TemplateDistribution::RequestTransactionData(
+                        m.into_static(),
+                    ))
+                }
+                TemplateDistribution::RequestTransactionDataError(m) => {
+                    AnyMessage::TemplateDistribution(
+                        TemplateDistribution::RequestTransactionDataError(m.into_static()),
+                    )
+                }
+                TemplateDistribution::RequestTransactionDataSuccess(m) => {
+                    AnyMessage::TemplateDistribution(
+                        TemplateDistribution::RequestTransactionDataSuccess(m.into_static()),
+                    )
+                }
+                TemplateDistribution::SetNewPrevHash(m) => AnyMessage::TemplateDistribution(
+                    TemplateDistribution::SetNewPrevHash(m.into_static()),
+                ),
+                TemplateDistribution::SubmitSolution(m) => AnyMessage::TemplateDistribution(
+                    TemplateDistribution::SubmitSolution(m.into_static()),
+                ),
+            },
+        }
+    }
+
+    async fn wait_for_client(client: SocketAddr) -> TcpStream {
+        let listner = TcpListener::bind(client)
+            .await
+            .expect("Impossible to listen on given address");
+        if let Ok((stream, _)) = listner.accept().await {
+            stream
+        } else {
+            panic!("Impossible to accept dowsntream connection")
+        }
+    }
+}
+
+// This implementation is used in order to check if a test has handled all messages sent by the
+// downstream and upstream roles. If not, the test will panic.
+//
+// This is useful to ensure that the test has checked all exchanged messages between the roles.
+impl Drop for Sniffer {
+    fn drop(&mut self) {
+        // Don't print backtrace on panic
+        std::panic::set_hook(Box::new(|_| {
+            println!();
+        }));
+        if !self.downstream_messages.is_empty() {
+            println!(
+                "You didn't handle all downstream messages: {:?}",
+                self.downstream_messages
+            );
+            panic!();
+        }
+        if !self.upstream_messages.is_empty() {
+            println!(
+                "You didn't handle all upstream messages: {:?}",
+                self.upstream_messages
+            );
+            panic!();
```

Comment:
Why do we need this? From what I understand, the sniffer acts as a proxy that intercepts messages between the roles. Once we start the roles and the sniffer, we can't really limit or control the flow of messages between them. The sniffer’s only job is to capture the messages exchanged between the two entities. However, we might only be interested in a specific prefix of those messages, not all of them. Right now, since we're just setting up the connection and not doing much else, we're only seeing a couple of messages. But once more roles are introduced, the amount of intercepted messages will explode, making it harder to manage. Currently, with the template provider, we're not doing much—just starting it up and mining the initial 16 blocks—so we're only getting basic connection setup messages. But as our testing becomes more complex, that’s going to change.I think we should drop these checks. Once we have the mocks set up, we'll be able to control the message flow more precisely, and in those scenarios, these checks will actually be helpful.

Reply:
let's start doing real tests

if we realize that this is becoming an issue we can remove it

Reply:
makes sense